IB/mlx4: set write-combining flag for userspace blueflame pages

Supported on i386 and x86_64 for now.

V2: Added support for PPC64, cleaned up warnings for unsupported platforms.

Signed-off-by: Michael S. Tsirkin <mst@dev.mellanox.co.il>
Signed-off-by: Jack Morgenstein <jackm@dev.mellanox.co.il>

Index: ofed_kernel/drivers/infiniband/hw/mlx4/Makefile
===================================================================
--- ofed_kernel.orig/drivers/infiniband/hw/mlx4/Makefile	2009-04-05 10:02:23.222077000 +0300
+++ ofed_kernel/drivers/infiniband/hw/mlx4/Makefile	2009-04-05 10:10:47.671938000 +0300
@@ -1,3 +1,4 @@
 obj-$(CONFIG_MLX4_INFINIBAND)	+= mlx4_ib.o
 
 mlx4_ib-y :=	ah.o cq.o doorbell.o mad.o main.o mr.o qp.o srq.o
+mlx4_ib-y +=	wc.o
Index: ofed_kernel/drivers/infiniband/hw/mlx4/main.c
===================================================================
--- ofed_kernel.orig/drivers/infiniband/hw/mlx4/main.c	2009-04-05 10:02:23.223082000 +0300
+++ ofed_kernel/drivers/infiniband/hw/mlx4/main.c	2009-04-05 10:11:29.782421000 +0300
@@ -43,6 +43,7 @@
 
 #include "mlx4_ib.h"
 #include "user.h"
+#include "wc.h"
 
 #define DRV_NAME	"mlx4_ib"
 #define DRV_VERSION	"1.0"
@@ -395,7 +396,7 @@ static int mlx4_ib_mmap(struct ib_uconte
 				       PAGE_SIZE, vma->vm_page_prot))
 			return -EAGAIN;
 	} else if (vma->vm_pgoff == 1 && dev->dev->caps.bf_reg_size != 0) {
-		vma->vm_page_prot = pgprot_writecombine(vma->vm_page_prot);
+		vma->vm_page_prot = pgprot_wc(vma->vm_page_prot);
 
 		if (io_remap_pfn_range(vma, vma->vm_start,
 				       to_mucontext(context)->uar.pfn +
Index: ofed_kernel/drivers/infiniband/hw/mlx4/wc.c
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ ofed_kernel/drivers/infiniband/hw/mlx4/wc.c	2009-04-05 10:13:42.026670000 +0300
@@ -0,0 +1,74 @@
+/*
+ * Copyright (c) 2006-2007 Mellanox Technologies. All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * OpenIB.org BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <linux/pci.h>
+#include "wc.h"
+
+#if defined(__i386__) || defined(__x86_64__)
+
+pgprot_t pgprot_wc(pgprot_t _prot)
+{
+	return pgprot_writecombine(_prot);
+}
+
+int mlx4_wc_enabled(void)
+{
+	return 1;
+}
+
+#elif defined(CONFIG_PPC64)
+
+pgprot_t pgprot_wc(pgprot_t _prot)
+{
+	return __pgprot((pgprot_val(_prot) | _PAGE_NO_CACHE) &
+				     ~(pgprot_t)_PAGE_GUARDED);
+}
+
+int mlx4_wc_enabled(void)
+{
+	return 1;
+}
+
+#else	/* !(defined(__i386__) || defined(__x86_64__)) */
+
+pgprot_t pgprot_wc(pgprot_t _prot)
+{
+	return pgprot_noncached(_prot);
+}
+
+int mlx4_wc_enabled(void)
+{
+	return 0;
+}
+
+#endif
+
Index: ofed_kernel/drivers/infiniband/hw/mlx4/wc.h
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ ofed_kernel/drivers/infiniband/hw/mlx4/wc.h	2009-04-05 10:10:47.686936000 +0300
@@ -0,0 +1,41 @@
+/*
+ * Copyright (c) 2006-2007 Mellanox Technologies. All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * OpenIB.org BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#ifndef mlx4_WC_H
+#define mlx4_WC_H
+
+#include <asm/pgtable.h>
+
+int mlx4_wc_enabled(void);
+pgprot_t pgprot_wc(pgprot_t _prot);
+
+#endif
